regression/cbmc/gcc_builtin_mul_overflow/main.c

```diff
@@ -0,0 +1,116 @@
+#include <assert.h>
+#include <limits.h>
+#include <stdint.h>
+
+#define A_VALUE_X_SO_THAT_X_TIMES_X_DOES_NOT_OVERFLOW(T)                       \
+  (((T)(1)) << (sizeof(T) * CHAR_BIT) / 2 - 1)
+
+void check_int(void)
+{
+  int result;
+  assert(!__builtin_smul_overflow(1, 1, &result));
+  assert(result == 1);
+  int const lt_isqrt_of_int_max =
+    A_VALUE_X_SO_THAT_X_TIMES_X_DOES_NOT_OVERFLOW(int);
+  assert(!__builtin_smul_overflow(
+    lt_isqrt_of_int_max, lt_isqrt_of_int_max, &result));
+  assert(result == lt_isqrt_of_int_max * lt_isqrt_of_int_max);
+  assert(__builtin_smul_overflow(
+    lt_isqrt_of_int_max << 1, lt_isqrt_of_int_max << 1, &result));
```

Comment:
These would cause quite a large overflow - what about also testing with `lt_isqrt_of_int_max + 1` to get closer to a boundary condition? Ditto for the tests for the other types.

Reply:
@chrisr-diffblue `(lt_sqrt_of_int_max+1)*(lt_sqrt_of_int_max+1)` wouldn’t actually overflow at all with the way it’s defined.